ngraph/frontend/frontend_manager/src/frontend_manager.cpp

```diff
@@ -0,0 +1,387 @@
+
+public:
+    Impl() { registerPlugins(); }
+
+    ~Impl() = default;
+
+    FrontEnd::Ptr loadByFramework(const std::string& framework, FrontEndCapFlags fec)
+    {
+        FRONT_END_ASSERT(m_factories.count(framework))
+        return m_factories[framework](fec);
+    }
+
+    std::vector<std::string> availableFrontEnds() const
+    {
+        std::vector<std::string> keys;
+
+        std::transform(
+            m_factories.begin(),
+            m_factories.end(),
+            std::back_inserter(keys),
+            [](const std::pair<std::string, FrontEndFactory>& item) { return item.first; });
+        return keys;
+    }
+
+    FrontEnd::Ptr loadByModel(const std::string& path, FrontEndCapFlags fec)
+    {
+        FRONT_END_NOT_IMPLEMENTED(loadByModel);
+    }
+
+    void registerFrontEnd(const std::string& name, FrontEndFactory creator)
+    {
+        m_factories.insert({name, creator});
+    }
+
+private:
+    void registerPlugins()
+    {
+        auto registerFromDir = [&](const std::string& dir) {
+            if (!dir.empty())
+            {
+                auto plugins = loadPlugins(dir);
+                for (auto& plugin : plugins)
+                {
+                    registerFrontEnd(plugin.m_pluginInfo.m_name, plugin.m_pluginInfo.m_creator);
+                    m_loadedLibs.push_back(std::move(plugin.m_libHandle));
+                }
+            }
+        };
+        std::string envPath = ngraph::getenv_string("OV_FRONTEND_PATH");
+        if (!envPath.empty())
+        {
+            auto start = 0u;
+            auto sepPos = envPath.find(PathSeparator, start);
+            while (sepPos != std::string::npos)
+            {
+                registerFromDir(envPath.substr(start, sepPos - start));
+                start = sepPos + 1;
+                sepPos = envPath.find(PathSeparator, start);
+            }
+            registerFromDir(envPath.substr(start, sepPos));
+        }
+        else
+        {
+            registerFromDir(".");
+        }
+    }
+};
+
+FrontEndManager::FrontEndManager()
+    : m_impl(new Impl())
+{
+}
+
+FrontEndManager::~FrontEndManager() = default;
+
+FrontEnd::Ptr FrontEndManager::load_by_framework(const std::string& framework, FrontEndCapFlags fec)
+{
+    return m_impl->loadByFramework(framework, fec);
+}
+
+FrontEnd::Ptr FrontEndManager::load_by_model(const std::string& path, FrontEndCapFlags fec)
+{
+    return m_impl->loadByModel(path, fec);
+}
+
+std::vector<std::string> FrontEndManager::get_available_front_ends() const
+{
+    return m_impl->availableFrontEnds();
+}
+
+void FrontEndManager::register_front_end(const std::string& name, FrontEndFactory creator)
+{
+    m_impl->registerFrontEnd(name, creator);
+}
+
+//----------- FrontEnd ---------------------------
+
+FrontEnd::FrontEnd() = default;
+
+FrontEnd::~FrontEnd() = default;
+
+InputModel::Ptr FrontEnd::load_from_file(const std::string& path) const
+{
+    FRONT_END_NOT_IMPLEMENTED(load_from_file);
```

Comment:
Looks like some methods (for example read) can be pure virtual.

Reply:
The idea is that frontends can implement only a sub-set of these methods, however we don't have API to check in runtime if some loaded frontend supports specific method or not.
Regarding load_from_file - this is also not a strict requirement for all frontends, some may potentially implement only 'load_from_files' or 'load_from_streams'. @slyalin, do you agree with this understanding?

Reply:
@mvafin is reworking these methods to something even more universal (as we discussed last week). But yes, some methods may not be implemented depending on FW -- not all FWs provides the same level of support. This is specific for methods that read models. For other methods that may not be implemented, it is defined by the capabilities level requested when FE is created. But currently I don't see value from requesting particular level of capabilities because there is a believe (that has to be proven though) that it doesn't give a real tool for binary size control (the difference in size is small between different capability levels). So even if the capability levels are not used when FE is created, not all methods have to be implemented. In this case FE is still usable, and depending on called methods it is allowed to throw an exception.